### PR TITLE
distrubuted create [#CLICKHOUSE-5]

### PR DESCRIPTION
**Pull request by artpaul**

### CODE REVIEW

File: dbms/include/DB/Parsers/ASTCreateQuery.h [hunk 1]
```diff
@@ -13,17 +13,18 @@ namespace DB
 class ASTCreateQuery : public IAST
 {
 public:
-	bool attach{false};	/// Запрос ATTACH TABLE, а не CREATE TABLE.
+	bool attach{false};	/// If true then it means what query is ATTACH TABLE.
```
Comment (alexey-milovidov) [inline dbms/include/DB/Parsers/ASTCreateQuery.h]: Little typo.

File: dbms/src/Interpreters/DDLWorker.cpp [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+{
+	extern const int UNKNOWN_ELEMENT_IN_CONFIG;
+	extern const int INVALID_CONFIG_PARAMETER;
+}
+
+namespace {
+
+/// Helper class which extracts from the ClickHouse configuration file
+/// the parameters we need for operating the resharding thread.
+class Arguments final
```
Comment (alexey-milovidov) [inline dbms/src/Interpreters/DDLWorker.cpp]: "resharding" was left from copy-paste.

File: dbms/src/Interpreters/DDLWorker.cpp [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+	{
+		Poco::Util::AbstractConfiguration::Keys keys;
+		config.keys(config_name, keys);
+
+		for (const auto & key : keys)
+		{
+			if (key == "task_queue_path")
+				task_queue_path = config.getString(config_name + "." + key);
+			else
+				throw Exception{"Unknown parameter in resharding configuration", ErrorCodes::UNKNOWN_ELEMENT_IN_CONFIG};
```
Comment (alexey-milovidov) [inline dbms/src/Interpreters/DDLWorker.cpp]: .

File: dbms/src/Interpreters/DDLWorker.cpp [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+		}
+
+		if (task_queue_path.empty())
+			throw Exception{"Resharding: missing parameter task_queue_path", ErrorCodes::INVALID_CONFIG_PARAMETER};
```
Comment (alexey-milovidov) [inline dbms/src/Interpreters/DDLWorker.cpp]: .

File: dbms/include/DB/Interpreters/DDLWorker.h [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+#include <atomic>
+#include <chrono>
+#include <condition_variable>
+#include <mutex>
+#include <thread>
+
+namespace DB
+{
+
+class DDLWorker
```
Comment (alexey-milovidov) [inline dbms/include/DB/Interpreters/DDLWorker.h]: Need comment about what it is.
What is the structure of nodes in ZooKeeper. How they are processed.
And so on.

File: dbms/src/Interpreters/DDLWorker.cpp [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+{
+	using namespace std::chrono_literals;
+
+	while (!stop_flag)
+	{
+		try
+		{
+			processTasks();
+		}
+		catch (const std::exception & ex)
```
Comment (alexey-milovidov) [inline dbms/src/Interpreters/DDLWorker.cpp]: Need to catch ... to extract more exception info.
See `tryLogCurrentException` across the code.

File: dbms/src/Interpreters/InterpreterCreateQuery.cpp [hunk 7]
```diff
@@ -56,6 +57,21 @@ namespace ErrorCodes
 	extern const int DUPLICATE_COLUMN;
 }
 
+static void ExecuteQuery(const Cluster::Address & addr, const std::string & query)
```
Comment (alexey-milovidov) [inline dbms/src/Interpreters/InterpreterCreateQuery.cpp]: Style.

File: dbms/src/Interpreters/InterpreterCreateQuery.cpp [hunk 8]
```diff
@@ -562,6 +578,62 @@ BlockIO InterpreterCreateQuery::createTable(ASTCreateQuery & create)
...
+	if (tmp.database.empty())
+		tmp.database = context.getCurrentDatabase();
+	return cloned;
+}
+
+void InterpreterCreateQuery::writeToZookeeper(ASTPtr query, const std::vector<Cluster::Address> & addrs)
+{
+	auto zookeeper = context.getZooKeeper();
+	ASTCreateQuery & create = typeid_cast<ASTCreateQuery &>(*query);
+	std::string table_name = create.database + "." + create.table;
```
Comment (alexey-milovidov) [inline dbms/src/Interpreters/InterpreterCreateQuery.cpp]: See `backQuoteIfNeed` for quoting of identifiers.

File: dbms/src/Interpreters/InterpreterCreateQuery.cpp [hunk 9]
```diff
@@ -562,6 +578,62 @@ BlockIO InterpreterCreateQuery::createTable(ASTCreateQuery & create)
...
+
+	for (const auto & addr : addrs)
+	{
+		const std::string & path =
+			"/clickhouse/task_queue/ddl/" +
+			addr.host_name +
+			"/create/" + table_name;
+
+		// TODO catch exceptions
```
Comment (alexey-milovidov) [inline dbms/src/Interpreters/InterpreterCreateQuery.cpp]: You could create all required nodes in single transaction atomically.

File: dbms/src/Interpreters/InterpreterCreateQuery.cpp [hunk 10]
```diff
@@ -562,6 +578,62 @@ BlockIO InterpreterCreateQuery::createTable(ASTCreateQuery & create)
...
+	{
+		for (const auto & addr : shard)
+		{
+			try
+			{
+				ExecuteQuery(addr, query);
+			}
+			catch (...)
+			{
+				failed.push_back(addr);
```
Comment (alexey-milovidov) [inline dbms/src/Interpreters/InterpreterCreateQuery.cpp]: Ok. But we also need to collect information from exceptions and then write log,
... and further we need to return a resultset for distributed create query, containing information, that following addresses was processed Ok, and following will be processed asynchronously.

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]

File: dbms/include/DB/Interpreters/InterpreterCreateQuery.h [hunk 11]
```diff
@@ -55,6 +56,10 @@ class InterpreterCreateQuery : public IInterpreter
 private:
 	void createDatabase(ASTCreateQuery & create);
 	BlockIO createTable(ASTCreateQuery & create);
+	BlockIO createTableOnCluster(ASTCreateQuery & create);
+
+	ASTPtr createQueryWithoutCluster(ASTCreateQuery & create) const;
+	void writeToZookeeper(ASTPtr query, const std::vector<Cluster::Address>& addrs);
```
Comment (alexey-milovidov) [inline dbms/include/DB/Interpreters/InterpreterCreateQuery.h]: Style.

File: dbms/src/Interpreters/InterpreterCreateQuery.cpp [hunk 12]
```diff
@@ -562,6 +578,62 @@ BlockIO InterpreterCreateQuery::createTable(ASTCreateQuery & create)
...
+	auto zookeeper = context.getZooKeeper();
+	ASTCreateQuery & create = typeid_cast<ASTCreateQuery &>(*query);
+	std::string table_name = create.database + "." + create.table;
+
+	for (const auto & addr : addrs)
+	{
+		const std::string & path =
+			"/clickhouse/task_queue/ddl/" +
+			addr.host_name +
+			"/create/" + table_name;
```
Comment (alexey-milovidov) [inline dbms/src/Interpreters/InterpreterCreateQuery.cpp]: It seems that `task_queue` is not actually a queue.
Queries are reordered. Need to mind that and don't forget to mention in doc.

File: dbms/src/Interpreters/DDLWorker.cpp [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+		{
+			LOG_ERROR(log, ex.what());
+		}
+
+		std::unique_lock<std::mutex> g(lock);
+		cond_var.wait_for(g, 10s);
+	}
```
Comment (alexey-milovidov) [inline dbms/src/Interpreters/DDLWorker.cpp]: I think, better to install watches.

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]

File: dbms/src/Interpreters/InterpreterCreateQuery.cpp [hunk 14]
```diff
@@ -562,6 +578,62 @@ BlockIO InterpreterCreateQuery::createTable(ASTCreateQuery & create)
...
+		{
+			try
+			{
+				ExecuteQuery(addr, query);
+			}
+			catch (...)
+			{
+				failed.push_back(addr);
+			}
+		}
```
Comment (alexey-milovidov) [inline dbms/src/Interpreters/InterpreterCreateQuery.cpp]: We must differentiate temporary failures (server unavailability, network issues) with incorrect queries.

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]

File: dbms/src/Parsers/formatAST.cpp [hunk 15]
```diff
@@ -1,9 +1,23 @@
...
 namespace DB
 {
 
+void formatAST(const IAST & ast, std::ostream & s, size_t indent, bool hilite, bool one_line)
+{
+	IAST::FormatSettings settings(s, hilite, one_line);
+	ast.format(settings);
+}
+
+std::string formatASTToString(const IAST & ast)
```
Comment (ludv1x) [inline dbms/src/Parsers/formatAST.cpp]: Maybe, existing [queryToString](https://github.com/yandex/ClickHouse/blob/master/dbms/include/DB/Parsers/queryToString.h) will be better.

**Comment by alexey-milovidov** [comment]
Waiting for new pull request.

**Pull request closed without merging**